Deprecate the Flash Middleware.

Edouard-chin opened the pull request:
Deprecate the Flash middleware:

- I originally created this PR to fix a bug where deleting the Flash Middleware wasn't doing anything, but after some discussions we realised that this middleware wasn't really one so we decided that it'd be better to just remove it (see https://github.com/rails/rails/pull/44588#issuecomment-1055845413)

  What that "flash.rb" file was doing was just monkeypatching the `ActionDispatch::Request` method once it was loaded, but it was very implicit. The goal of this PR is to:

 - Make things more explicit when monkeypatching
 - Move everything from the `middleware/flash.rb` file in a more appropriate locations (because it's not a middleware)
 - Deprecate the `ActionDispatch::Flash` middleware. We need to deprecate it because that middleware was documented as being part of the default middleware stack and thus users could legitimately reference it in their app when inserting other middlewares (i.e. `config.middleware.insert_after(ActionDispatch::Flash, MyMiddleware)`)

The first commit is the largest one and just moves everything from the `flash.rb` file. The main changes are in the other commits

byroot commented:
Hum, should removing the middleware prevent that module from being prepended though? Maybe it would be cleaner to just have flash access raise some error if the middleware was removed?

Edouard-chin commented:
Hello @byroot 👋 

> Hum, should removing the middleware prevent that module from being prepended though? [...]

The implementation of that Flash middleware is curious, but it's not a middleware, it's a monkeypatch disguised as a middleware (it doesn't even have a `call` method and instantiating it returns the previous middleware from the stack, not an instance of itself)https://github.com/rails/rails/blob/ad2529be4b5942b6f38ab83b9796e1976685edca/actionpack/lib/action_dispatch/middleware/flash.rb#L293

What I'm trying to say is that once the once the file is loaded (and monkeypatch applied), that middleware is useless, so the only logical things to do when one remove it is to prevent the patch.

Maybe the cleanest solution would be to delete it entirely and just do a `monkey_patch if uses_flash?`, though it's a change that is a bit more involved.

byroot commented:
That's curious indeed. I'll do some git spelunking to see if there's a reason for this.

byroot commented:
Ok, so it has been make that way since https://github.com/rails/rails/commit/ca324a0d9080660dc9c1803f90976cecb81b5f58

I think we should just deprecate that middleware, even if it's an empty middleware it still end up on the stack so it's an extra call, and an extra frame for no good reason.

Edouard-chin commented:
Yeah agreed to delete and deprecate, pretty sure there are bunch of apps that did `middleware.use Blabla, after: Flash`

kwstannard commented on railties/test/application/middleware/flash_test.rb:
```diff
@@ -0,0 +1,29 @@
+
+module ApplicationTests
+  class FlashTest < ActiveSupport::TestCase
+    include ActiveSupport::Testing::Isolation
```
Hey! So, if I am understanding this file correctly, this will create a fork prior to loading `action_dispatch/middleware/flash.rb` so that the Request class is clean for the tests in this class? That is pretty snazzy.

Edouard-chin replied:
> so that the Request class is clean for the tests in this class

That's correct :)

tenderlove commented:
> Yeah agreed to deprecate, pretty sure there are bunch of apps that did middleware.use Blabla, after: Flash

Yes, that was the problem.  I'd like it if we just removed the middleware though.  Maybe we can make it work but add a deprecation notice if someone passes `Flash` in to the middleware builder?

> Hum, should removing the middleware prevent that module from being prepended though? Maybe it would be cleaner to just have flash access raise some error if the middleware was removed?

+1 on this.  Removing the monkey patch seems confusing and I'd rather calling the flash methods raise an exception.  We could probably detect whether flash is supported or not at the time the request object is allocated, and configure the request with a "null flash" or something (but that's just an idea)

kwstannard commented:
> +1 on this. Removing the monkey patch seems confusing and I'd rather calling the flash methods raise an exception. We could probably detect whether flash is supported or not at the time the request object is allocated, and configure the request with a "null flash" or something (but that's just an idea)

You might be interested in this pr: https://github.com/rails/rails/pull/43760/files

Edouard-chin committed:
Move everything related to the Flash middleware:

- Since the Flash Middleware is no longer a middleware (see
  https://github.com/rails/rails/pull/44588#issuecomment-1055845413
  for the reason) it doesn't make sense to have it in the Middleware
  folder.

  Also decided to split all nested classes inside separate files,
  make things easier to read.

  This commit is in prepration to deprecated the `Flash::Middleware`
  constant.

Edouard-chin committed:
Deprecate the `ActionDispatch::Flash` middleware:

- This middleware was not doing anything since https://github.com/rails/rails/commit/ca324a0d9080660dc9c1803f90976cecb81b5f58

  We are planning to remove it but we need to deprecate it first,
  because it is documented as part of the default middleware stack,
  so users could legitimatelly reference it to insert other middleware
  in the stack (in example by calling
  `config.middleware.insert_after(ActionDispatch::Flash, MyMiddleware)).

Edouard-chin committed:
Modify all references to `ActionDispatch::Flash`:

- Now using `Request::Flash::FlashHash`. Ultimatelly, Rails itself
  no longer uses `ActionDispatch::Flash` so the only way for that
  constant to be loaded is if an application explicitely references
  it and thus would trigger the deprecation message.

Edouard-chin committed:
Remove very old code to support rails 3.2:

- The FlashHash used to be marshalized in the session prior to Rails 4
  This code was added for backward compatibility, I think it's time
  to get rid of it.

Edouard-chin committed:
Remove reference to the flash middleware from the docs

Edouard-chin commented:
Modified this PR and its description to deprecate this middleware. The line changes reflected are because of moving stuff around in the first commit, the others commit are the most interesting.

Edouard-chin commented on railties/lib/rails/application/default_middleware_stack.rb:
```diff
@@ -71,7 +71,6 @@ def build_stack
           end
 
           unless config.api_only
-            middleware.use ::ActionDispatch::Flash
```
Hmm, actually had a brain fart. I need to leave it for now, otherwise it will break app if one do `config.middleware.insert_after(Flash, MyMiddleware)`. Will modify

kwstannard commented:
The latest changes look like they bring in what I was doing in my PR. I can close mine.

rafaelfranca commented on guides/source/api_app.md:
```diff
@@ -371,7 +371,6 @@ API application, especially if one of your API clients is the browser:
 
 - `Rack::MethodOverride`
 - `ActionDispatch::Cookies`
-- `ActionDispatch::Flash`
```
We now need to document how add flash to API only apps.

rafaelfranca commented on actionview/test/abstract_unit.rb:
```diff
@@ -101,6 +101,8 @@ def config
 
 class ActionDispatch::IntegrationTest < ActiveSupport::TestCase
   def self.build_app(routes = nil)
+    ActionDispatch::Request::Flash.use!
```
Isn't this going to prepend the module more than once?

Edouard-chin replied:
Should be fine:

```
irb(main):001:0> module T; end
irb(main):002:0> 10.times { NilClass.prepend(t) }
irb(main):004:0> NilClass.ancestors

=> [T, NilClass, Object, PP::ObjectMixin, Kernel, BasicObject]
```

rafaelfranca commented on actionpack/lib/action_dispatch/request/flash.rb:
```diff
@@ -0,0 +1,52 @@
+
+      extend self
+
+      def use! # :nodoc:
```
This will need to be doced because we want API apps to be able to add flash if they want to.

Edouard-chin committed:
fixup! Deprecate the `ActionDispatch::Flash` middleware:

Edouard-chin committed:
PR REVIEW: Take in consideration API-only apps:

- Added docs on how to opt-in for the flash feature now
  that the middleware is depracated.

- Added another deprecation and test to make sure API-only apps
  that previously opted-in to flash will see the deprecation.
  There is now 2 distincts deprecation:

  ```ruby
  middleware.insert_after(ActionDispatch::Flash, MyMiddleware)
  # Deprecated with no replacement

  middleware.use(ActionDispatch::Flash)
  # Deprecated with a message saying how to migrate
  ```

skipkayhil commented on actionpack/lib/action_dispatch/request/flash.rb:
```diff
@@ -46,7 +46,13 @@ module Flash
+      # Rails applications *not* in API mode are automatically configured, calling this method is not required.
+      # If your application is a Rails API-only one, you can refer to the guide to know how to opt-in.
+      #
+      # If you are using ActionPack as a standalone gem, call this method during the boot process of your application.
```
```suggestion
      # If you are using Action Pack as a standalone gem, call this method during the boot process of your application.
```

skipkayhil commented on actionpack/lib/action_dispatch/request/flash.rb:
```diff
@@ -46,7 +46,13 @@ module Flash
       extend self
 
-      def use! # :nodoc:
+      # Method to call (`ActionDsiaptch::Request::Flash.use!`) to enable the flash feature in your application.
```
```suggestion
      # Method to call (`ActionDispatch::Request::Flash.use!`) to enable the flash feature in your application.
```

skipkayhil commented on actionpack/lib/action_dispatch/request/flash.rb:
```diff
@@ -46,7 +46,13 @@ module Flash
+      # Method to call (`ActionDsiaptch::Request::Flash.use!`) to enable the flash feature in your application.
+      #
+      # Rails applications *not* in API mode are automatically configured, calling this method is not required.
+      # If your application is a Rails API-only one, you can refer to the guide to know how to opt-in.
```
I think we should either explain this here or link to the guide

Edouard-chin replied:
Thanks, make sense 👍 . I added some doc

Edouard-chin committed:
fixup! PR REVIEW: Take in consideration API-only apps:

Edouard-chin commented:
THe rails repo introduced a max number of PR opened per user, so doing a spring cleaning of my stale PRs.

kwstannard commented:
Can we re-open this then? https://github.com/rails/rails/pull/43760

Edouard-chin commented:
Sure! Feel free to reopen your PR !
